File: console/components/shared/levels/levelTier.ts
/**
 * Level tier helpers. Levels come in five tiers of three:
 * 1-3 Iron, 4-6 Jade, 7-9 Amethyst, 10-12 Gold, 13-15 Ruby.
 * Each level's `color` field normally holds its tier hex; we prefer it when valid
 * and fall back to the tier colour derived from the level number.
 */

export type LevelTierName = "Iron" | "Jade" | "Amethyst" | "Gold" | "Ruby";

export interface LevelTier {
  name: LevelTierName;
  color: string;
  minLevel: number;
  maxLevel: number;
}

export const LEVEL_TIERS: readonly LevelTier[] = [
  { name: "Iron", color: "#8A96A8", minLevel: 1, maxLevel: 3 },
  { name: "Jade", color: "#2AA876", minLevel: 4, maxLevel: 6 },
  { name: "Amethyst", color: "#8B5CF6", minLevel: 7, maxLevel: 9 },
  { name: "Gold", color: "#F0B429", minLevel: 10, maxLevel: 12 },
  { name: "Ruby", color: "#E23D4A", minLevel: 13, maxLevel: 15 },
];

const HEX_COLOR_PATTERN = /^#(?:[0-9a-f]{3}|[0-9a-f]{6})$/i;

export function isHexColor(value: string | null | undefined): value is string {
  return typeof value === "string" && HEX_COLOR_PATTERN.test(value.trim());
}

/** Tier for a level number. Out-of-range values clamp to the first/last tier. */
export function getLevelTier(level: number): LevelTier {
  const first = LEVEL_TIERS[0];
  const last = LEVEL_TIERS[LEVEL_TIERS.length - 1];
  if (!Number.isFinite(level) || level < first.minLevel) return first;
  if (level > last.maxLevel) return last;
  return (
    LEVEL_TIERS.find((tier) => level >= tier.minLevel && level <= tier.maxLevel) ??
    first
  );
}

export interface LevelColorSource {
  level: number;
  color?: string | null;
}

/** Colour for a level, preferring its own `color` field when it is a valid hex. */
export function getLevelColor(source: LevelColorSource): string {
  if (isHexColor(source.color)) return source.color.trim();
  return getLevelTier(source.level).color;
}

export function formatTierRange(tier: LevelTier): string {
  return `Levels ${tier.minLevel}–${tier.maxLevel}`;
}

/** Translucent version of a colour for tints (uses CSS color-mix). */
export function tintColor(color: string, percent: number): string {
  return `color-mix(in srgb, ${color} ${percent}%, transparent)`;
}

export interface LevelTierGroup<T> {
  tier: LevelTier;
  levels: T[];
}

/**
 * Groups levels by tier, in tier order. Within a tier, levels are sorted by
 * `sortOrder` (when present) then by level number. Empty tiers are omitted.
 */
export function groupLevelsByTier<
  T extends { level: number; sortOrder?: number },
>(levels: readonly T[]): LevelTierGroup<T>[] {
  const byTier = new Map<LevelTierName, T[]>();
  for (const level of levels) {
    const tier = getLevelTier(level.level);
    const bucket = byTier.get(tier.name) ?? [];
    bucket.push(level);
    byTier.set(tier.name, bucket);
  }

  return LEVEL_TIERS.flatMap((tier) => {
    const bucket = byTier.get(tier.name);
    if (!bucket || bucket.length === 0) return [];
    const sorted = [...bucket].sort(
      (a, b) =>
        (a.sortOrder ?? 0) - (b.sortOrder ?? 0) || a.level - b.level,
    );
    return [{ tier, levels: sorted }];
  });
}

File: mobile/lib/features/levels/domain/level_tier.dart
import 'dart:ui';

import 'package:blocnet/features/levels/data/models/user_level_model.dart';

/// One of the five progression tiers, each spanning three levels.
///
/// The palette mirrors the backend seed (`seed.levels.ts`) and is used as a
/// fallback when a level arrives without a `color`. Use [colorFor] to resolve
/// the accent for a concrete level, which prefers the server-provided value.
class LevelTier {
  const LevelTier._({
    required this.index,
    required this.name,
    required this.shape,
    required this.minLevel,
    required this.maxLevel,
    required this.fallbackColor,
  });

  /// Zero-based tier position (Iron = 0 … Ruby = 4).
  final int index;
  final String name;

  /// Badge silhouette used by the bundled artwork for this tier.
  final String shape;
  final int minLevel;
  final int maxLevel;
  final Color fallbackColor;

  static const int levelsPerTier = 3;

  static const LevelTier iron = LevelTier._(
    index: 0,
    name: 'Iron',
    shape: 'Circle',
    minLevel: 1,
    maxLevel: 3,
    fallbackColor: Color(0xFF8A96A8),
  );

  static const LevelTier jade = LevelTier._(
    index: 1,
    name: 'Jade',
    shape: 'Hexagon',
    minLevel: 4,
    maxLevel: 6,
    fallbackColor: Color(0xFF2AA876),
  );

  static const LevelTier amethyst = LevelTier._(
    index: 2,
    name: 'Amethyst',
    shape: 'Shield',
    minLevel: 7,
    maxLevel: 9,
    fallbackColor: Color(0xFF8B5CF6),
  );

  static const LevelTier gold = LevelTier._(
    index: 3,
    name: 'Gold',
    shape: 'Octagon',
    minLevel: 10,
    maxLevel: 12,
    fallbackColor: Color(0xFFF0B429),
  );

  static const LevelTier ruby = LevelTier._(
    index: 4,
    name: 'Ruby',
    shape: 'Seal',
    minLevel: 13,
    maxLevel: 15,
    fallbackColor: Color(0xFFE23D4A),
  );

  /// All tiers in progression order.
  static const List<LevelTier> all = [iron, jade, amethyst, gold, ruby];

  /// Resolves the tier for a level number. Levels below 1 map to Iron and
  /// levels above 15 map to Ruby so an unexpected server value never throws.
  static LevelTier forLevel(int level) {
    final raw = (level - 1) ~/ levelsPerTier;
    final clamped = raw.clamp(0, all.length - 1);
    return all[clamped];
  }

  /// Human readable range, e.g. `Levels 4–6`.
  String get rangeLabel => 'Levels $minLevel–$maxLevel';

  bool contains(int level) => level >= minLevel && level <= maxLevel;

  bool get isFirst => index == 0;
  bool get isLast => index == all.length - 1;

  /// Accent colour for [level]: the server colour when it parses, otherwise
  /// the tier palette.
  Color colorFor(UserLevelModel? level) {
    return tryParseHexColor(level?.color) ?? fallbackColor;
  }
}

/// Convenience accessors on a level.
extension UserLevelTier on UserLevelModel {
  LevelTier get tier => LevelTier.forLevel(level);

  /// Tier accent for this level, preferring the server-provided colour.
  Color get tierColor => tier.colorFor(this);
}

/// A tier together with the levels that belong to it, in level order.
class LevelTierSection {
  const LevelTierSection({
    required this.tier,
    required this.levels,
    required this.color,
  });

  final LevelTier tier;
  final List<UserLevelModel> levels;

  /// Accent for the whole section, taken from the first level that carries a
  /// valid colour, otherwise the tier palette.
  final Color color;

  bool get isEmpty => levels.isEmpty;
}

/// Groups [levels] into ordered tier sections. Tiers without any level are
/// omitted so a partial server response never renders empty headers.
List<LevelTierSection> groupLevelsByTier(List<UserLevelModel> levels) {
  final buckets = <int, List<UserLevelModel>>{};
  for (final level in levels) {
    buckets
        .putIfAbsent(LevelTier.forLevel(level.level).index, () => [])
        .add(level);
  }

  final sections = <LevelTierSection>[];
  for (final tier in LevelTier.all) {
    final bucket = buckets[tier.index];
    if (bucket == null || bucket.isEmpty) continue;
    bucket.sort((a, b) => a.level.compareTo(b.level));

    Color? color;
    for (final level in bucket) {
      color = tryParseHexColor(level.color);
      if (color != null) break;
    }

    sections.add(LevelTierSection(
      tier: tier,
      levels: List.unmodifiable(bucket),
      color: color ?? tier.fallbackColor,
    ));
  }
  return sections;
}

/// Parses `#RRGGBB` / `RRGGBB` / `#AARRGGBB`; returns `null` when invalid.
Color? tryParseHexColor(String? hex) {
  if (hex == null) return null;
  final cleaned = hex.trim().replaceFirst('#', '');
  if (cleaned.length != 6 && cleaned.length != 8) return null;
  final value = int.tryParse(cleaned, radix: 16);
  if (value == null) return null;
  return cleaned.length == 6 ? Color(0xFF000000 | value) : Color(value);
}

/// Black or white, whichever reads better on [background].
Color foregroundOn(Color background) {
  return background.computeLuminance() > 0.4
      ? const Color(0xFF09090B)
      : const Color(0xFFFFFFFF);
}
